text/0864-deprecate-ember-a.md

```diff
@@ -0,0 +1,94 @@
+---
+stage: accepted
+start-date: 2022-11-22T00:00:00.000Z
+release-date:
+release-versions:
+teams:
+  - data
+  - framework
+  - learning
+  - typescript
+prs:
+  accepted: https://github.com/emberjs/rfcs/pull/864
+project-link:
+---
+
+# Deprecate Ember.A()
+
+## Summary
+
+Remove `Ember.A()` as the functionality it provides is no longer needed in the post octane reactivity system and with the removal of array prototype extensions in #848.
+
+## Motivation
+
+`Ember.A()` provides a way for addon authors or application owners who disable array prototype extensions to add reactivity to arrays (through `pushObject()`, `removeObject()`) and to access prototype extension convenience methods such as `filterBy()` on an any array-like object. Both of these paradigms have been deprecated in Ember and `Ember.A()` should follow suit as it can break things in unexpected ways and interferes with progress in [Ember](https://github.com/emberjs/ember.js/blob/4339725976299b24c69fb9dfbf13d18bf9917130/packages/@ember/-internals/utils/lib/ember-array.ts) and [Ember Data](https://github.com/emberjs/data/blob/47a71ca1538ba9e2d7dfa01bf048a2db897bdf5f/packages/store/addon/-private/record-arrays/identifier-array.ts#L381-L401) where special care must be taken to prevent or compensate for usage.
+
+Along with it's usefulness `Ember.A()` has a significant API gotcha in that it doesn't return a new instance of the array-like object it is passed, it modifies that object and returns is such that `A(array) === array` this can create the dreaded spooky-action-at-a-distance effect where suddenly an object which was `Array` can become an `Ember.NativeArray`. An example of this is using `{{includes "Zoey" this.mascots}}` from [ember-composable-helpers](https://github.com/DockYard/ember-composable-helpers#includes) will modify `this.mascots` and add `lastObject` to the API. Re-ordering the template code at a later date will result in a mysterious failure that can be extremely frustrating to understand.
+
+```js
+export default class WeirdArray extends Component {
+  mascots = ['Tomster', 'Zoey'];
+
+  get lastMascot() {
+    return this.mascots.lastObject;
+  }
+
+  <template>
+    {{#if (includes "Zoey" this.mascots)}}
+      Zoey Rules!
+    {{/if}}
+    Last Mascot: {{this.lastMascot}}
+  </template>
+}
+```
+
+Deprecating `Ember.A()` will provide a signal to the addon community to move away from both array prototype extensions and array reactivity which requires using the Emberism `pushObject`/`removeObject` to track updates. 
+
+Now that array prototype extensions have been deprecated it is quite possible that usage of `Ember.A()` will increase significantly as a defensive coding strategy to maintain access to the family of methods Ember adds to all arrays. This should be immediately discouraged as `Ember.A()` is not a long term viable solution and other better alternatives exist.
```

Comment:
This is a very good reason to deprecate Ember.A, ASAP.

Reply:
Credit to @runspired, I stole this from https://github.com/emberjs/rfcs/pull/848#issuecomment-1235883364